src/CloudEvents/V1/CloudEvent.php

```diff
@@ -145,4 +164,47 @@ public function setTime(?DateTimeInterface $time): CloudEvent
 
         return $this;
     }
+
+    /**
+     * @return array<string,bool|int|string>
+     */
+    public function getExtensions(): array
+    {
+        return $this->extensions;
+    }
+
+    /**
+     * @param bool|int|string|null $value
+     */
+    public function setExtension(string $attribute, $value): CloudEvent
+    {
+        if (\preg_match('/^[a-z0-9]+$/', $attribute) !== 1) {
+            throw new ValueError(
+                \sprintf('%s(): Argument #1 ($attribute) must match the regex [a-z0-9]+, %s given', __METHOD__, $attribute)
+            );
+        }
+
+        if (in_array($attribute, self::RESERVED_ATTRIBUTES, true)) {
+            throw new ValueError(
+                \sprintf('%s(): Argument #1 ($attribute) must not be a reserved attribute, %s given', __METHOD__, $attribute)
+            );
+        }
+
+        $type = \get_debug_type($value);
+        $types = ['bool', 'int', 'string', 'null'];
+
+        if (!in_array($type, $types, true)) {
+            throw new TypeError(
+                \sprintf('%s(): Argument #2 ($value) must be of type %s, %s given', __METHOD__, implode('|', $types), $type)
```

Comment:
Notice `DateTimeImmutable` is not supported. This is inline with how the Java SDK. The reason for that behaviour is that there's no way to tell the difference between a string that looks like a datetime and an actual datetime. We thus leave it to the user to decode and encode datetimes (possibly using the stuff in `CloudEvents\Serializers\Formatters`) before providing using them as extensions.